File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/src/sphere.cpp
/*
** EPITECH PROJECT, 2026
** snd_year_projects
** File description:
** sphere
*/

#include "../include/sphere.hpp"
using namespace RayTracer;

Sphere::Sphere(const Point3D& givenCenter, double givenRadius)
    : center(givenCenter), radius(givenRadius)
{}

bool Sphere::hits(const Ray& ray)
{
    Vector3D origin_center = ray.origin - center;
    double a = ray.direction.dot(ray.direction);
    double b = 2 * ray.direction.dot(origin_center);
    double c  = origin_center.dot(origin_center) - pow(radius, 2);

    double discriminant = pow(b, 2) - (4 * a * c);
    if (discriminant >= 0)
        return true;
    return false;
}

File: second_year/projects/G-OOP-400-COT-4-1-bsraytracer-13/rust/src/sphere.rs
//
// EPITECH PROJECT, 2026
// snd_year_projects
// File description:
// sphere
//

use crate::Point3D;
use crate::Ray;

pub struct Sphere {
    pub center: Point3D,
    pub radius: f32, 
}

impl Sphere {
    pub fn hits(&self, ray: Ray) -> bool {
        let origin_center = ray.origin - self.center;
        let a = ray.direction.dot(ray.direction);
        let b = 2.0 * ray.direction.dot(origin_center);
        let c = origin_center.dot(origin_center) - self.radius.powf(2.0);
        let determinant = b.powf(2.0) - (4.0 * a * c);

        if determinant >= 0.0 {
            return true;
        }
        return false;
    }
}
